### PR TITLE
ch2

### PR DESCRIPTION
**Pull request by 0girls1cup**
### Solutions for Chapter 2
another one bites the dust
cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +336,9 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList f t 
+  | f >= 0 && t >= 0 = drop f . take (succ t)
```
Comment (chshersh) [inline src/Chapter2.hs]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -349,7 +351,10 @@ Implement a function that returns only the first half of a given list.
 "b"
 -}
 -- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l = go l 0 where
+  go (_:_:r) n = go r (succ n)
+  go _       n = take n l
```
Comment (chshersh) [inline src/Chapter2.hs]: Interesting implementation :brain: 
Though, you implement an even smarter solution by traversing the same list twice: with step 1 and 2 at the same time. With this approach, you will be able to return list elements immediately, and your solution will work on infinite lists as well.

File: src/Chapter2.hs [hunk 3]
```diff
@@ -535,7 +542,7 @@ For example, we can patch the previous function to count the number of
 steps we need to take in order to reduce the number to zero.
 
 🤔 Blitz question: can you guess what this number represents?
-
+<guess>recursion depth?</guess>
```
Comment (chshersh) [inline src/Chapter2.hs:544]: Yes, but what does this recursion depth represents :thinking:

File: src/Chapter2.hs [hunk 4]
```diff
@@ -621,7 +629,9 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven (e:_:r) = e : takeEven r
+takeEven r       = r
```
Comment (chshersh) [inline src/Chapter2.hs]: I would recommend here to cover the remaining cases explicitly to take advantage of GHC pattern matching checker coverage :+1: 
```suggestion
takeEven [x]       = [x]
takeEven []        = []
```

File: src/Chapter2.hs [hunk 5]
```diff
@@ -842,7 +855,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate d l 
+  | d >= 0    = let n = length l in take n $ drop d $ cycle l
```
Comment (chshersh) [inline src/Chapter2.hs]: That's a correct solution! Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -858,7 +874,8 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind = foldl (flip (:)) []
```
Comment (chshersh) [inline src/Chapter2.hs:879]: Excellent implementation :gem:

**Review by chshersh — APPROVED** [review]
Amazing work on Chapter Two tasks! :clap: 
You nailed the advanced task :hammer: 

I left some comments with general suggestions and some feedback on your solutions :slightly_smiling_face:

**Comment by 0girls1cup** [comment]
made some changes @chshersh

**Pull request merged**